### ISSUE TITLE
srcID can't be an Array

### ISSUE THREAD

**Issue by sensestage** [ModalityTeam/Modality-toolkit#222]
This is an issue in MIDIMKtlDevice, where this can happen when there is more than one port to a device; so we probably have to take the first one by default, and make explicit port selection possible again.

```
(
// make a MIDIOut for the tests in the examples
Platform.case(
    \linux, {
        ~midiOut= MIDIOut.new(0); // = SC's first midi output
        ~midiOut.connect(1); // = connect to SC's first midi input
    },
    \osx, {
        ~midiOut = MIDIOut.newByName( "IAC Driver", "IAC Bus 1" );
    }
    /*,
    \windows, {
        // please let us know what this should be!
    }
    */
);
)

MIDIdef.noteOn( \test, { arg ...args; [ 1, args ].postln; }, nil, nil, 8388611 );
MIDIdef.noteOn( \test2, { arg ...args; [ 2, args ].postln; }, nil, nil, [ 8388611 ] );
```

**Comment by adcxyz** [comment]
for inputs, several inputs should be possible that can be done by giving the MIDIFunc nil as srcID, 
and doing the matching for the array inside it. for outputs, we do have to know which one is the right one, 
and put it in the desc file.

**Comment by sensestage** [comment]
But we want to be specific where the MIDI comes in; it might be that two different MIDI ports of the same device send the same commands on the same channel, but on a different port.

A more common case may be that we have several devices, which send the same kind of data via MIDI, and having srcID as nil would also catch the info from other devices.

**Comment by sensestage** [comment]
Our original idea, btw, was to just have one srcID and one destID per MKtl description and then make a combined MKtl if a device has multiple, similarly as we would do for a device that has a MIDI and an HID interface.

**Comment by sensestage** [comment]
See: #162, #21, #129, #178

**Comment by adcxyz** [comment]
Yes I remember, and I wonder whether it is better to keep that, or simpler to absorb multiple MIDI ports into a single desc file (semantically, because it is a single physical device). 
Will try to rework MIDIExplorer soon to see better what actual multiport devices use the different ports for, in order to understand how best to support that.

**Comment by LFSaw** [comment]
In the to-be-merged MKtl_refactor_test2 branch, we fixed this issue by allowing to set the following parameters:

```
m = MKtl('a', "devicewithmultipleports")
-> does not create responders and reports about that
m.listenTo(<index>) -- creates responders for that index (and throws away previously assigned ones)
m.sendTo(<index>) -- creates Output for that index (and throws away previously assigned ones)
```

This is currently only implemented for MIDI, since we assume this to be a MIDI-only issue.
In the future, it should be possible to combine those MKtls into a composite MKtl.

**Comment by sensestage** [comment]
similar case happens for OSC; the application could be sending from a different port than it listens to.

**Comment by sensestage** [comment]
In the solution, is it possible to define the portIndex in the description file? I think this is essential - we want the device to be defined in the description file - not in user code.

**Comment by LFSaw** [comment]
> similar case happens for OSC; the application could be sending from a different port than it listens to.

send and receive are completely separated.

> In the solution, is it possible to define the portIndex in the description file? I think this is essential - we want the device to be defined in the description file - not in user code.

not for now but it is certainly possible to add a flag for that.
However, I'd like to first make the merge/move of the refactor2 branch to master before investing more time and effort in new ideas and features.
My approach would be to make small steps instead of large ones and try to keep everyone involved.

**Comment by adcxyz** [comment]
in the sektch I just committed, we tried putting all elements for all ports in one big desc. 
this means you choose by index from the full desc of the device in user code.

we also saw that there are other complexities that will make that approach hard to finish, 
so we will go back to single-port desc files. some design questions here remain, 
mainly how to organize the files; will ask those in separate mail.

best a
